### PR TITLE
Documentation Cleanup and restructering: Kernel

### PR DESCRIPTION
**Pull request by nashif**
- Flatten kernel documentation and remove too many intermediate pages and incorporate API documentation in the main documentation.
- Added extension to support linking to git files.
- Change to CSS to removetop border from doxygen documentation

Related to #12671

### CODE REVIEW

File: doc/extensions/zephyr/link-roles.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    else:
+        return 'master'
+
+
+def setup(app):
+    rev = get_github_rev()
+
+    # links to files or folders on the GitHub
+    baseurl = 'https://github.com/zephyrproject-rtos/zephyr'
+    app.add_role('zephyr_file', autolink('{}/blob/{}/%s'.format(baseurl, rev)))
```
Comment (dbkinder) [inline doc/extensions/zephyr/link-roles.py]: We need to document these two new roles in our doc guidelines (doc/documentation/doc-guidelines.rst).  Let me know if you'd like me to do that in another PR.

File: doc/extensions/zephyr/link-roles.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+from __future__ import print_function
+from __future__ import unicode_literals
+import re
+import os
+from docutils import nodes
+from local_util import run_cmd_get_output
+
+
+def get_github_rev():
+    tag = run_cmd_get_output('git describe --exact-match')
```
Comment (dbkinder) [inline doc/extensions/zephyr/link-roles.py]: When on an untagged branch (such as master) this writes out something like: ``fatal: no tag exactly matches '9d36fe0e2c7b4fdcb998d5ded5443ecffbcb19c9'`` I suspect to stderr so it shows up on the output of the ``make htmldocs``.  I think you can hide that stderr output ...

File: doc/kernel/data_passing/fifos.rst [hunk 3]
```diff
@@ -154,17 +154,10 @@ Configuration Options
...
-The following fifo APIs are provided by :file:`kernel.h`:
-
-* :c:macro:`K_FIFO_DEFINE`
-* :cpp:func:`k_fifo_init()`
-* :cpp:func:`k_fifo_alloc_put()`
-* :cpp:func:`k_fifo_put()`
-* :cpp:func:`k_fifo_put_list()`
-* :cpp:func:`k_fifo_put_slist()`
-* :cpp:func:`k_fifo_get()`
+.. doxygengroup:: fifo_apis
```
Comment (dbkinder) [inline doc/kernel/data_passing/fifos.rst]: Nice. Dumping the group is much better than listing individual entities.  Not sure why we were doing that before.

File: doc/kernel/kernel.rst [hunk 4]
```diff
@@ -25,10 +25,25 @@ include: fitness wearables, smart watches, and IoT wireless gateways.
 .. toctree::
    :maxdepth: 2
 
-   threads/threads.rst
-   timing/timing.rst
-   memory/memory.rst
-   synchronization/synchronization.rst
-   data_passing/data_passing.rst
+   threads/index.rst
```
Comment (dbkinder) [inline doc/kernel/kernel.rst]: If you wanted to group these with a heading for each group, you can use multiple toctree directives...

File: doc/static/zephyr-custom.css [hunk 5]
```diff
@@ -134,12 +135,13 @@ th,td {
 }
 
 /* doxygenXX item color tweaks, light blue background with dark blue top border */
-.rst-content dl:not(.docutils) dl dt {
+.rst-content dl:not(.docutils) dl dt, dl:not(.docutils) dt {
   background: #e7f2fa !important;
-  border-top: solid 3px #2980B9 !important;
+  border-top: none !important;
```
Comment (dbkinder) [inline doc/static/zephyr-custom.css]: You don't like that top border, eh?  OK.

File: doc/kernel/other/ring_buffers.rst [hunk 6]
```diff
@@ -303,22 +303,10 @@ operations on the ring buffer's memory.  For example:
...
-* :cpp:func:`ring_buf_put()`
-* :cpp:func:`ring_buf_put_claim()`
-* :cpp:func:`ring_buf_put_finish()`
-* :cpp:func:`ring_buf_get()`
-* :cpp:func:`ring_buf_get_claim()`
-* :cpp:func:`ring_buf_get_finish()`
+API Reference
+*************
+
+The following ring buffer APIs are provided by :zephyr_file:`include/ring_buffer.h`:
```
Comment (dbkinder) [inline doc/kernel/other/ring_buffers.rst]: This is rather cool!  I can do a general scan/replace after this PR gets merged for most of the :file:`xxx` references in other documents.  (I have a script I run occasionally to check that :file:`xxx` references are to files that exist.)

File: doc/extensions/local_util.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+# limitations under the License.
+
+from __future__ import unicode_literals
+from io import open
+import os
+import shutil
+
+
+def run_cmd_get_output(cmd):
+    return os.popen(cmd).read().strip()
```
Comment (dbkinder) [inline doc/extensions/local_util.py]: I think if you use subprocess.popen (https://docs.python.org/3/library/subprocess.html#popen-constructor) you can direct stderr to /dev/null so the command is quieter for our cases.

**Review by dbkinder — APPROVED** [review]
Looks great.  I left a few comments...

File: doc/conf.py [hunk 8]
```diff
@@ -46,7 +46,7 @@
     'sphinx.ext.autodoc',
     'zephyr.application',
     'zephyr.html_redirects',
-    'only.eager_only'
+    'zephyr.link-roles'
```
[the remark this answers is not in the document]
Comment (nashif) [reply doc/conf.py]: I dropped it by mistake during a rebase, will re-add

File: doc/extensions/zephyr/link-roles.py (hunk 1 above)
Comment (nashif) [reply doc/extensions/zephyr/link-roles.py]: can be added later when you do a mass search/replace :)

**Review by dbkinder — APPROVED** [review]
Nice.
